pkg/controller/admissionchecks/multikueue/fswatch.go

```diff
@@ -0,0 +1,223 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package multikueue
+
+import (
+	"context"
+	"errors"
+	"path"
+	"sync"
+
+	"github.com/fsnotify/fsnotify"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/set"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+
+	kueuealpha "sigs.k8s.io/kueue/apis/kueue/v1alpha1"
+)
+
+var (
+	errNotStarted = errors.New("not started")
+)
+
+type KubeConfigFSWatcher struct {
+	watcher       *fsnotify.Watcher
+	lock          sync.RWMutex
+	clusterToFile map[string]string
+	// a single file can be potentially used by multiple clusters
+	fileToClusters   map[string]set.Set[string]
+	parentDirToFiles map[string]set.Set[string]
+	reconcile        chan event.GenericEvent
+}
+
+var _ manager.Runnable = (*KubeConfigFSWatcher)(nil)
+
+func newKubeConfigFSWatcher() *KubeConfigFSWatcher {
+	return &KubeConfigFSWatcher{
+		clusterToFile:    map[string]string{},
+		fileToClusters:   map[string]set.Set[string]{},
+		parentDirToFiles: map[string]set.Set[string]{},
+		reconcile:        make(chan event.GenericEvent),
+	}
+}
+
+func (w *KubeConfigFSWatcher) Start(ctx context.Context) error {
+	w.lock.Lock()
+	defer w.lock.Unlock()
+
+	var err error
+	w.watcher, err = fsnotify.NewWatcher()
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		log := ctrl.LoggerFrom(ctx).WithName("kubeconfig-fs-watcher")
+		defer func() {
+			err := w.watcher.Close()
+			if err != nil {
+				log.Error(err, "Closing kubeconfigs FS Watcher")
+			}
+		}()
+		watchedEvents := fsnotify.Write | fsnotify.Create | fsnotify.Remove | fsnotify.Rename
+		for {
+			select {
+			case ev := <-w.watcher.Events:
+				log.V(5).Info("Got event", "name", ev.Name, "op", ev.Op)
+				switch {
+				case (ev.Op & watchedEvents) != 0:
+					w.notifyPathWrite(ev.Name)
+				default:
+					log.V(2).Info("Ignore event", "name", ev.Name, "op", ev.Op)
+
+				}
+			case err := <-w.watcher.Errors:
+				log.Error(err, "Kubeconfigs FS Watch")
+			case <-ctx.Done():
+				log.V(2).Info("End kubeconfigs FS Watch")
+				return
+			}
+		}
+
+	}()
+	return nil
+}
+
+func (w *KubeConfigFSWatcher) notifyPathWrite(path string) {
+	w.lock.RLock()
+	clusters := w.fileToClusters[path].UnsortedList()
+	w.lock.RUnlock()
```

Comment:
I guess it would be safer to extract to a helper function `clustersForPath(path)` and unlock in defer.